File: app/src/main/java/eu/geekhome/rest/configurable/ConfigurableController.java
package eu.geekhome.rest.configurable;

import com.geekhome.common.configurable.Configurable;
import eu.geekhome.rest.PluginsManager;

import javax.inject.Inject;
import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import java.util.List;
import java.util.stream.Collectors;

@Path("configurables")
public class ConfigurableController {

    private PluginsManager _pluginsManager;
    private ConfigurableDtoMapper _configurableDtoMapper;

    @Inject
    public ConfigurableController(PluginsManager pluginsManager, ConfigurableDtoMapper configurableDtoMapper) {
        _pluginsManager = pluginsManager;
        _configurableDtoMapper = configurableDtoMapper;
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    public List<ConfigurableDto> getConfigurables(@Context HttpServletRequest request) {

        List<Configurable> allConfigurables = _pluginsManager
                .getConfigurables();

        return _pluginsManager
                .getConfigurables()
                .stream()
                .filter((x) -> x.getParent() == null)
                .map((x) -> _configurableDtoMapper.map(x, allConfigurables))
                .collect(Collectors.toList());
    }
}

File: app-server/src/main/java/eu/automateeverything/rest/configurables/ConfigurableController.kt
package eu.automateeverything.rest.configurables

import eu.automateeverything.domain.extensibility.PluginsCoordinator
import jakarta.inject.Inject
import jakarta.ws.rs.GET
import jakarta.ws.rs.Produces
import eu.automateeverything.data.configurables.ConfigurableDto
import jakarta.servlet.http.HttpServletRequest
import java.util.stream.Collectors
import jakarta.ws.rs.Path
import jakarta.ws.rs.core.Context
import jakarta.ws.rs.core.MediaType

@Path("configurables")
class ConfigurableController @Inject constructor(
    private val pluginsCoordinator: PluginsCoordinator,
    private val configurableDtoMapper: ConfigurableDtoMapper
) {
    @GET
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    fun getConfigurables(): List<ConfigurableDto> {
        return pluginsCoordinator
            .configurables
            .stream()
            .map { configurableDtoMapper.map(it) }
            .collect(Collectors.toList())
    }
}